### ISSUE TITLE
Optimize load times and memory footprints of form entry with entities

### ISSUE THREAD

**Issue by seadowg** [getodk/collect#5623]
~~Blocked by #5694~~

This just for entity secondary instances **not all secondary instances**.

## Optimize load times

When a form is loaded in Collect calculates are run as part of "initialization" (when `FormDef#initialize` is called). This means that the complexity of calculate expressions affects the for load time. Recent changes to JavaRosa have introduced lazy indexing and caching to some calculate expressions on secondary instances which lets them run faster after this initialization step, but all of these solutions are "in memory" so the speed-up doesn't persist between form entry sessions.

We want to make changes to Collect that result in faster subsequent "loads" (probably really initialization) for forms that contain the following restricted set of calculate expressions on external secondary instances that compare nodes in the instance to something else of the following form:

-  [x] `/<name> = <"absolute" expression>`
- [x] `/<entity property> = <"absolute" expression>`

The "absolute" part can be an absolute path expression, a string literal or a numeric literal.

To measure this, we should use timings on a known slow device (like @lognaturel's trusty test phone). We're aiming for speed-ups (at least for subsequent loads) of around the order of an O(n) to O(1) complexity here.

## Decrease memory footprint

Currently, external secondary instances are loaded into memory with the rest of the form. This is potentially problematic for larger datasets (we often see them exceed hundreds of thousands of records). Along with speeding up load times, we want to reduce the memory footprint of form entry for forms like this (at least on subsequent loads). This again only needs to apply to forms with our above restricted set of calculate expressions.

To measure this, we should use Android Studio's profiler with any emulator and monitor the total memory footprint of a loaded form with a larger dataset (100,000+ records). We should measure the same form with and without the dataset so that we have a rough idea of the size of the dataset in memory. We are aiming to see the total memory footprint drop so that the dataset's footprint is virtually gone (in at least subsequent loads).

**Comment by seadowg** [comment]
@lognaturel as described in the issue here, I've realized that our recent optimization work didn't touch external secondary instances when used as itemsets. My thinking is that most forms of this kind would include a select on the instance that's later used to calculate things in the form (the latter of which we're optimizing and planning to improve on here). I still think this work is worth doing, but it feels like we need to apply the "predicate filter chain" (or something similar) to itemsets in JavaRosa (and then expand on it in Collect) to really get the speed-ups and memory drops we want?

**Comment by seadowg** [comment]
>  as described in the issue here, I've realized that our recent optimization work didn't touch external secondary instances when used as itemsets.

First step to clearing "needs discussion" here should be proving/disproving this statement.

**Comment by seadowg** [comment]
Digging into the above question, it seems that we do have some level of caching applied to choice filters on selects. It is implemented in a completely different way though, the new caching only affects evaluations that go through `TriggerableDag` where as choice lists are computed on demand, and the evaluation is mostly owned by `ItemsetBinding` (it doesn't touch `TriggerableDag`). To illustrate this, here's one passing and one failing test:

**Passing:**
```java
Scenario scenario = Scenario.init("Some form", html(
    head(
        title("Some form"),
        model(
            mainInstance(t("data id=\"some-form\"",
                t("choice"),
                t("select")
            )),
            instance("instance",
                item("a", "A"),
                item("b", "B")
            ),
            bind("/data/choice").type("string"),
            bind("/data/select").type("string")
        )
    ),
    body(
        input("/data/choice"),
        select1Dynamic("/data/select", "instance('instance')/root/item[value = /data/choice]")
    )
));

int evaluations = Measure.withMeasure(asList("PredicateEvaluation", "IndexEvaluation"), () -> {
    scenario.answer("/data/choice", "a");

    scenario.choicesOf("/data/select");
    scenario.choicesOf("/data/select");
});

// Check that we do less than (size of secondary instance) * (number of choice lookups)
assertThat(evaluations, lessThan(4));
```

**Failing:**

```java
Scenario scenario = Scenario.init("Some form", html(
    head(
        title("Some form"),
        model(
            mainInstance(t("data id=\"some-form\"",
                t("choice"),
                t("select")
            )),
            instance("instance",
                item("a", "A"),
                item("b", "B")
            ),
            bind("/data/choice").type("string"),
            bind("/data/select").type("string")
        )
    ),
    body(
        input("/data/choice"),
        select1Dynamic("/data/select", "instance('instance')/root/item[value = /data/choice]")
    )
));

int evaluations = Measure.withMeasure(asList("PredicateEvaluation", "IndexEvaluation"), () -> {
    scenario.answer("/data/choice", "a");
    scenario.choicesOf("/data/select");

    scenario.answer("/data/choice", "b");
    scenario.choicesOf("/data/select");
});

// Check that we do less than (size of secondary instance) * (number of choice lookups)
assertThat(evaluations, lessThan(4));
```

The first test passes because we cache choice list evaluations so that repeated look-ups (with the exact same literal expression) don't need to do anything. This doesn't use our predicate filter concept, so there's currently no way for a client (like Collect) to hook into the choice list evaluation like there is for all other expressions. This isn't the end of the world, but it would make our goal of reducing the memory footprint moot as we'd always load the full instance into memory to calculate choices whether we optimize how other expressions work or not.

I think the next step is to spike on replacing the current caching implementation for choice lists with the same `PredicateFilter` chain that the `TriggerableDag` uses with an aim of getting the failing test also passing. That would mean we could update `FormEntryController#addPredicateFilter` to modify the filter chain used on choice list evaluations as well (enabling us to add a `PredicateFilter` that reduces the memory footprint).

**Comment by seadowg** [comment]
It does look like we can use our new `PredicateFilter` approach to get the second test passing (specifically using `IndexPredicateFilter`) without having to remove the existing caching. I just spiked this out and manually added the filter to the `EvaluationContext` that `ItemsetBinding` uses, but I'm pretty sure we'd be able to rework it so that `TriggerableDag` and `ItemsetBinding` share a chain of `PredicateFilter` instances that can be customized externally with `FormEntryController#addPredicateFilter`.

Now that's worked out, I've run into another issue: if we want to reduce the memory footprint as detailed above (so that sessions after the first only use the memory needed to load the same form without the external instance), we'll run into trouble due to the choice list caching in `ItemsetBinding` as it will keep the last evaluated choice list (for each question as far as I understand) in memory. For a `select1` backed by an instance (without filtering), this would mean the whole instance would be retained in memory by `ItemsetBinding` (in `cachedFilteredChoiceList`) regardless of what we do while evaluating the choice list. I think our solution to this issue can get around this by reworking the existing `ItemsetBinding` caching implementation so that it's a `PredicateFilter` that's further down the chain than a theoretical one Collect adds to avoid loading anything into memory (discussed briefly in "Notes"). 

Regardless of the complexity involved in reworking the existing `ItemsetBinding` caching, it feels to me like the two goals ("Optimize load times" and "Decrease memory footprint") are feasible. Let me know if you agree/disagree @lognaturel.

**Comment by seadowg** [comment]
>  think the next step is to spike on replacing the current caching implementation for choice lists with the same PredicateFilter chain that the TriggerableDag uses with an aim of getting the failing test also passing. That would mean we could update FormEntryController#addPredicateFilter to modify the filter chain used on choice list evaluations as well (enabling us to add a PredicateFilter that reduces the memory footprint).

I've pulled this out to a [new issue](https://github.com/getodk/collect/issues/5694) that's not blocked by the above discussion.

**Comment by seadowg** [comment]
I've started investigating how to approach this, and I've sketched out a `FilterStrategy` that **almost** compiles:

```kotlin
class InstanceDataFilterStrategy(private val instanceDataAdapter: InstanceDataAdapter) : FilterStrategy {
    override fun filter(
        sourceInstance: DataInstance<*>,
        nodeSet: TreeReference,
        predicate: XPathExpression,
        children: MutableList<TreeReference>,
        evaluationContext: EvaluationContext,
        next: Supplier<MutableList<TreeReference>>
    ): List<TreeReference> {
        val compareToNodeExpression = CompareToNodeExpression.parse(predicate)

        // Query adapter for elements with matching child
        val field = compareToNodeExpression!!.nodeSide.steps[0].name.name
        val value = compareToNodeExpression.evalContextSide(sourceInstance, evaluationContext)
        val elements = instanceDataAdapter.query(field, value as String)

        // Populate data in ExternalDataInstance
        (sourceInstance as ExternalDataInstance).populate(elements)

        // Return refs of elements
        return elements.map { it.ref }
    }
}
```

This is very simplistic and assumes our nodeset expression is just `instance('<instance name>')/root/item[<child>=<path>]` (which is the most important kind of expression to optimize anyway). It also assumes we've built some mechanism to allow `ExternalDataInstance` to only partially populate instances so they don't include child values (most likely using a custom `FileInstanceParser`) so we're not storing all the instance data in memory. The basic approach here is the grab the parts of the expression we need (the child we want to match against and the value to use), query against a new `InstanceDataAdapter` interface to get the matching `TreeElement`s, populate those elements into the `ExternalDataInstance` (so that they're ready for `#resolveReference`) and then return the `TreeReference`s for the elements.


The new interface and a very basic implementation for entities would look like this:

```kotlin
interface InstanceDataAdapter {
    fun query(child: String, value: String): List<TreeElement>
}

class EntitiesInstanceDataAdapter(val entitiesRepository: EntitiesRepository) : InstanceDataAdapter {
    override fun query(child: String, value: String): List<TreeElement> {
        val entities = entitiesRepository.getAllByProperty(child, value)
        return entities.map { convertToElement(it) }
    }

    private fun convertToElement(entity: Entity): TreeElement {
        val name = TreeElement(EntityItemElement.ID)
        name.value = StringData(entity.id)

        val label = TreeElement(EntityItemElement.LABEL)
        label.value = StringData(entity.label)

        val version = TreeElement(EntityItemElement.VERSION)
        version.value = StringData(entity.version.toString())

        val item = TreeElement("item", multiplicity)
        item.addChild(name)
        item.addChild(label)
        item.addChild(version)

        entity.properties.forEach { nameAndValue ->
            item.addChild(
                TreeElement(nameAndValue.first).also {
                    it.value = StringData(nameAndValue.second)
                }
            )
        }

        return item
    }
}
```

The things here we currently don't have implemented the `#getAllByProperty` method, and you'll also a way to define `multiplicity`. The former isn't hard to imagine, but the latter presents some questions. 

We could store mutliplicity in `EntitiesRepository`, or we could create a second data store that maps from entity ID. I don't really see a disadvantage to the former here: we could also use it when creating the instance from the `EntitiesRepository` for forms (currently we just use the index while looping through). Does that also make sense to you @lognaturel? 

I can imagine that for normal instances (which we don't need to tackle yet), we just build another implementation of `InstanceDataAdapter` that's backed by a DB representation of the csv instance (that either gets put together on the fly or during form download like we do with entities).

EDIT: I guess another thing to point out here is that we'd need some way to fall back to loading the full instance if we encounter a non-filtered nodeset expression like `instance('<instance name>')/root/item` (possibly by always lazy parsing external instances at `#resolveReference`). In that scenario, I think we'd always need to load the full instance into memory as we currently have no way of "streaming" nodesets to the UI.

**Comment by seadowg** [comment]
Digging in even more with a spike of implementing partial parsing (as mentioned above), it'd also be ideal to be able to store multiplicity in `EntitiesRepostitory` for that to work so that there's a way to know which entity a "partial" element is referring to without having to keep the ID in memory. I think making that change is going to be a good first step.

**Comment by seadowg** [comment]
Just going to solidify the steps I want to take here:

1. Add multiplicity to entities
2. Add "partial" elements concept (for `DataInstance`) to allow `FileInstanceParser` to provide partially parsed instances (with a plugin for providing a way to "fallback" if JavaRosa needs those elements fully populated)
3. Add filter strategy that queries `EntitiesRepository` for `=` expressions
4. Preload matched elements from the filter strategy into the relevant instance all at once so we don't use a fallback for each individual element
5. Preload all elements all at once for relevant instances when skipping predicates (and loading all items in the instance)

After 3, we'll technically have completed the "Decrease memory footprint", but will most likely have made everything slower as we'll be probably be loading instance elements individually from the `EntitiesRepository` the first time JavaRosa needs them. This would mean that whatever I/O `EntitiesRepository` would be done N times for a filter that matches N items (although this would only happen once per form entry session). 4 and 5 will fix that by allowing us to use the data we would already have in the filter strategy (4) and by performing a more intelligent load of everything when there are no filters (5).

The "Optimize load times" goal might actually be hard to accomplish in real world terms without first reimplementing `EntitiesRepository` (as we'll probably end up pushing our complexity there). That's being tracked and discussed in https://github.com/getodk/collect/issues/6012. I think for this issue we should measure complexity in accesses of `EntitiesRepository`.

**Comment by seadowg** [comment]
@lognaturel I've updated the issue description again so we just account for `name` and properties for issue, and that we just handle the `=` operator. We'll most likely want to expand on that, but I want to make sure we don't get bogged down in this over other entity work (like getting in sync with the new entity spec or fleshing out proper storage).
